crates/data_components/src/dynamodb/request_builder.rs

```diff
@@ -0,0 +1,1462 @@
+            key_values.extend(filter_values);
+            query_params = query_params.filter_expression(filter_str);
+        }
+
+        if !key_values.is_empty() {
+            query_params = query_params.expression_attribute_values(key_values);
+        }
+
+        if let Some(proj) = projection {
+            query_params = query_params.projection_expression(proj);
+        }
+
+        if !attribute_names.is_empty() {
+            query_params = query_params.expression_attribute_names(attribute_names);
+        }
+
+        let query = query_params
+            .build()
+            .map_err(|e| DataFusionError::Execution(format!("{e}").to_string()))?;
+        Ok(DynamoDBRequestPlan::Query(query))
+    }
+
+    fn build_scan_request(
+        &self,
+        filters: &[Expr],
+        projection: Option<String>,
+        attribute_names: HashMap<String, String>,
+        limit: Option<i32>,
+    ) -> DataFusionResult<DynamoDBRequestPlan> {
+        let mut scan_params =
+            ScanParamsBuilder::default().table_name(self.schema.table_name().to_string());
+
+        if filters.is_empty() {
+            // We only apply limit when there's no filter_expression.
+            // This is because in DynamoDB filter is applied before filters.
+            // As such it may end returning fewer records than we want.
+            if let Some(l) = limit {
+                scan_params = scan_params.limit(l);
+            }
+        } else {
+            let (filter_str, attribute_values) = self.build_filter_expression(filters);
+            if !filter_str.is_empty() {
+                scan_params = scan_params.filter_expression(filter_str);
+                scan_params = scan_params.expression_attribute_values(attribute_values);
+            }
+        }
+
+        if let Some(proj) = projection {
+            scan_params = scan_params.projection_expression(proj);
+        }
+
+        if !attribute_names.is_empty() {
+            scan_params = scan_params.expression_attribute_names(attribute_names);
+        }
+
+        let scan = scan_params
+            .build()
+            .map_err(|e| DataFusionError::Execution(format!("{e}").to_string()))?;
+        Ok(DynamoDBRequestPlan::Scan(scan))
+    }
+
+    fn extract_attribute_names(&self, filters: &[Expr]) -> HashMap<String, String> {
+        let mut attribute_names = HashMap::new();
+        for expr in filters {
+            self.extract_columns_from_expr(expr, &mut attribute_names);
+        }
+        attribute_names
+    }
+
+    fn extract_columns_from_expr(
+        &self,
+        expr: &Expr,
+        attribute_names: &mut HashMap<String, String>,
+    ) {
+        match expr {
+            Expr::Column(col) => {
+                if let Some(alias) = self.schema.get_column_alias(col.name()) {
+                    attribute_names.insert(alias.to_string(), col.name.to_string());
+                }
+            }
+            Expr::BinaryExpr(BinaryExpr { left, right, .. }) => {
```

Comment:
As an example of why the expr.apply is useful, this BinaryExpr match goes away completely.